object_database/web/ActiveWebService.py

```diff
@@ -99,12 +226,75 @@ def doWork(self, shouldStop):
 
     def configureApp(self):
         instanceName = self.serviceObject.name
-        self.app.route("/")(lambda: redirect("/services"))
-        self.app.route('/content/<path:path>')(self.sendContent)
-        self.app.route('/services')(self.sendPage)
-        self.app.route('/services/<path:path>')(self.sendPage)
-        self.sockets.route("/socket/<path:path>")(self.mainSocket)
+        self.app.config['SECRET_KEY'] = os.environ.get('SECRET_KEY') or genToken()
 
+        self.app.add_url_rule('/', endpoint=None, view_func=lambda: redirect("/services"))
+        self.app.add_url_rule('/content/<path:path>', endpoint=None, view_func=self.sendContent)
+        self.app.add_url_rule('/services', endpoint=None, view_func=self.sendPage)
+        self.app.add_url_rule('/services/<path:path>', endpoint=None, view_func=self.sendPage)
+        self.app.add_url_rule('/login', endpoint=None, view_func=self.login, methods=['GET', 'POST'])
+        self.app.add_url_rule('/logout', endpoint=None, view_func=self.logout)
+        self.sockets.add_url_rule('/socket/<path:path>', None, self.mainSocket)
+
+    def login_user(self, username, remember=False):
+        with self.db.transaction():
+            users = User.lookupAll(username=username)
+
+            if len(users) == 0:
```

Comment:
this lookup-or-create pattern ought to have a 'revisionConflictRetry' around it.

Reply:
:+1: